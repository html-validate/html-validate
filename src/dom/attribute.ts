import { Location } from "../context";
import { DynamicValue } from "./dynamic-value";

interface StaticAttribute {
	readonly value: string | null;
}

interface DynamicAttribute {
	readonly value: DynamicValue;
}

/**
 * Type guard for [[Attribute]] testing if the value is static (missing value is
 * considered static as it is a fixed known value).
 *
 * @public
 */
export function isStaticAttribute(attr: Attribute | null): attr is Attribute & StaticAttribute {
	return Boolean(attr && attr.isStatic);
}

/**
 * Type guard for [[Attribute]] testing if the value is dynamic.
 *
 * @public
 */
export function isDynamicAttribute(attr: Attribute | null): attr is Attribute & DynamicAttribute {
	return Boolean(attr && attr.isDynamic);
}

/**
 * DOM Attribute.
 *
 * Represents a HTML attribute. Can contain either a fixed static value or a
 * placeholder for dynamic values (e.g. interpolated).
 */
export class Attribute {
	/** Attribute name */
	public readonly key: string;
	public readonly value: string | DynamicValue | null;
	public readonly keyLocation: Location;
	public readonly valueLocation: Location | null;
	public readonly originalAttribute?: string;

	/**
	 * @param key - Attribute name.
	 * @param value - Attribute value. Set to `null` for boolean attributes.
	 * @param keyLocation - Source location of attribute name.
	 * @param valueLocation - Source location of attribute value.
	 * @param originalAttribute - If this attribute was dynamically added via a
	 * transformation (e.g. vuejs `:id` generating the `id` attribute) this
	 * parameter should be set to the attribute name of the source attribute (`:id`).
	 */
	public constructor(
		key: string,
		value: null | string | DynamicValue | null,
		keyLocation: Location,
		valueLocation: Location | null,
		originalAttribute?: string
	) {
		this.key = key;
		this.value = value;
		this.keyLocation = keyLocation;
		this.valueLocation = valueLocation;
		this.originalAttribute = originalAttribute;

		/* force undefined to null */
		if (typeof this.value === "undefined") {
			this.value = null;
		}
	}

	/**
	 * Flag set to true if the attribute value is static.
	 */
	public get isStatic(): boolean {
		return !this.isDynamic;
	}

	/**
	 * Flag set to true if the attribute value is dynamic.
	 */
	public get isDynamic(): boolean {
		return this.value instanceof DynamicValue;
	}

	/**
	 * Test attribute value.
	 *
	 * @param pattern - Pattern to match value against. RegExp or a string (===)
	 * @param dynamicMatches - If true `DynamicValue` will always match, if false
	 * it never matches.
	 * @returns `true` if attribute value matches pattern.
	 */
	public valueMatches(pattern: RegExp, dynamicMatches?: boolean): boolean;
	public valueMatches(pattern: string, dynamicMatches?: boolean): boolean;
	public valueMatches(pattern: RegExp | string, dynamicMatches: boolean = true): boolean {
		if (this.value === null) {
			return false;
		}

		/* dynamic values matches everything */
		if (this.value instanceof DynamicValue) {
			return dynamicMatches;
		}

		/* test value against pattern */
		if (pattern instanceof RegExp) {
			return this.value.match(pattern) !== null;
		} else {
			return this.value === pattern;
		}
	}
}
