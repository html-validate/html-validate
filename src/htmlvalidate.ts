import { Config, ConfigLoader } from './config';
import { Engine, TokenDump, EventDump } from './engine';
import { Parser } from './parser';
import { Report } from './reporter';
import { Source } from './context';

class HtmlValidate {
	private globalConfig: Config;

	constructor(options?: any){
		const defaults = Config.empty();
		this.globalConfig = defaults.merge(options ? Config.fromObject(options) : Config.defaultConfig());
	}

	/**
	 * Parse HTML from string.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateString(str: string): Report {
		const source = {
			data: str,
			filename: 'inline',
			line: 1,
			column: 1,
		};
		return this.validateSource(source);
	}

	/**
	 * Parse HTML from source.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateSource(source: Source): Report {
		const config = this.getConfigFor('inline');
		const engine = new Engine(config, Parser);
		return engine.lint([source]);
	}

	/**
	 * Parse HTML from file.
	 *
	 * @param filename {string} - Filename to read and parse.
	 * @return {object} - Report output.
	 */
	public validateFile(filename: string): Report {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.lint(source);
	}

	public dumpTokens(filename: string): TokenDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTokens(source);
	}

	public dumpEvents(filename: string): EventDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpEvents(source);
	}

	public dumpTree(filename: string): string[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTree(source);
	}

	/**
	 * Get parser for given filename.
	 */
	public getParserFor(source: Source): Parser {
		const config = this.getConfigFor(source.filename);
		return new Parser(config);
	}

	/**
	 * Get configuration for given filename.
	 */
	public getConfigFor(filename: string): Config {
		const loader = new ConfigLoader();
		const config = loader.fromTarget(filename);
		const merged = this.globalConfig.merge(config);
		merged.init();
		return merged;
	}
}

export default HtmlValidate;
