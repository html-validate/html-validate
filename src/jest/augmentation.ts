/** module augmentation: jest.d.ts */
import { type ConfigData } from "../config";
import { type Message } from "../message";

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace -- module augmentation */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/no-empty-object-type, @typescript-eslint/no-unused-vars -- to match jest declaration */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toHaveError(error: Partial<Message>): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(errors: Array<[string, string] | Record<string, unknown>>): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 * @param filename - Optional filename used when matching transformer and
			 * loading configuration.
			 */
			toHTMLValidate(filename?: string): R;
			toHTMLValidate(config: ConfigData, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, config: ConfigData, filename?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with snapshot.
			 */
			toMatchCodeframe(snapshot?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with inline snapshot.
			 */
			toMatchInlineCodeframe(snapshot?: string): R;
		}
	}
}
/** module augmentation end */
