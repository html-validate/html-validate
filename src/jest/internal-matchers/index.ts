export { toBeToken } from "./to-be-token";
