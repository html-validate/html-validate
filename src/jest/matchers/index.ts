export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
