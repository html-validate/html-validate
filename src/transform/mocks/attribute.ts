import { DynamicValue } from "../../dom";
import { type AttributeData } from "../../parser";

export function* processAttribute(attr: AttributeData): Iterable<AttributeData> {
	/* handle foo="{{ bar }}" as "foo" with a dynamic value (interpolated) */
	if (typeof attr.value === "string" && /{{.*}}/.exec(attr.value)) {
		yield {
			...attr,
			value: new DynamicValue(attr.value),
		};
		return;
	}

	/* passthru original attribute */
	yield attr;

	/* handle "dynamic-foo" as alias for "foo" with dynamic value */
	if (attr.key.startsWith("dynamic-")) {
		yield {
			...attr,
			key: attr.key.replace("dynamic-", ""),
			value: new DynamicValue(attr.value as string),
			originalAttribute: attr.key,
		};
	}
}
