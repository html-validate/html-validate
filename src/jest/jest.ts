import "./augmentation";

import {
	toBeValid,
	toBeInvalid,
	toHTMLValidate,
	toHaveError,
	toHaveErrors,
	toMatchCodeframe,
	toMatchInlineCodeframe,
} from "./matchers";
import { diff } from "./utils";

expect.extend({
	toBeValid: toBeValid(),
	toBeInvalid: toBeInvalid(),
	toHTMLValidate: toHTMLValidate(expect, diff),
	toHaveError: toHaveError(expect, diff),
	toHaveErrors: toHaveErrors(expect, diff),
	toMatchCodeframe: toMatchCodeframe(),
	toMatchInlineCodeframe: toMatchInlineCodeframe(),
});
