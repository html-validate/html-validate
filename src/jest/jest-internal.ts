import "./jest";

import { Location } from "../context";
import { TokenType } from "../lexer";
import { toBeToken } from "./internal-matchers";

interface TokenMatcher {
	type: TokenType;
	location?: Partial<Location>;
	data?: any;
}

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-unused-vars */
		interface Matchers<R, T = {}> {
			toBeToken(expected: TokenMatcher): R;
		}
	}
}

expect.extend({
	toBeToken,
});
