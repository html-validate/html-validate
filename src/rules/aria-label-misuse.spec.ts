import { HtmlValidate } from "../htmlvalidate";
import "../jest";
import { processAttribute } from "../transform/mocks/attribute";

describe("rule aria-label-misuse", () => {
	let htmlvalidate: HtmlValidate;

	beforeAll(() => {
		htmlvalidate = new HtmlValidate({
			elements: [
				"html5",
				{
					"custom-allowed": {
						attributes: {
							"aria-label": {},
						},
					},
					"custom-disallowed": {
						attributes: {},
					},
				},
			],
			rules: { "aria-label-misuse": ["error"] },
		});
	});

	it("should not report for element without aria-label", async () => {
		expect.assertions(1);
		const markup = /* HTML */ ` <p></p> `;
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should not report for element with empty aria-label", async () => {
		expect.assertions(1);
		const markup = /* HTML */ ` <p aria-label=""></p> `;
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should not report for element with boolean aria-label", async () => {
		expect.assertions(1);
		const markup = /* HTML */ ` <p aria-label></p> `;
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should not report for element without meta", async () => {
		expect.assertions(1);
		const markup = /* HTML */ ` <custom-element aria-label="foobar"></custom-element> `;
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should not report for element with role", async () => {
		expect.assertions(1);
		const markup = /* HTML */ ` <p aria-label="foobar" role="widget"></p> `;
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	describe("should not report error for", () => {
		it.each`
			markup                                         | description
			${'<button aria-label="foobar"></button>'}     | ${"Interactive elements"}
			${'<input type="text" aria-label="foobar">'}   | ${"Labelable elements"}
			${'<main aria-label="foobar"></main>'}         | ${"Landmark elements"}
			${'<p role="widget" aria-label="foobar">'}     | ${'[role=".."]'}
			${'<p tabindex="0" aria-label="foobar"></p>'}  | ${"[tabindex]"}
			${'<area aria-label="foobar"></area>'}         | ${"<area>"}
			${'<dialog aria-label="foobar"></dialog>'}     | ${"<dialog>"}
			${'<form aria-label="foobar"></form>'}         | ${"<form>"}
			${'<fieldset aria-label="foobar"></fieldset>'} | ${"<fieldset>"}
			${'<iframe aria-label="foobar"></iframe>'}     | ${"<iframe>"}
			${'<img aria-label="foobar">'}                 | ${"<img>"}
			${'<figure aria-label="foobar"></figure>'}     | ${"<figure>"}
			${'<summary aria-label="foobar"></summary>'}   | ${"<summary>"}
			${'<table aria-label="foobar"></table>'}       | ${"<table>"}
			${'<td aria-label="foobar"></td>'}             | ${"<td>"}
			${'<th aria-label="foobar"></th>'}             | ${"<th>"}
			${'<any aria-label="foobar"></any>'}           | ${"<any>"}
		`("$description", async ({ markup }: { markup: string }) => {
			expect.assertions(1);
			const report = await htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});
	});

	it("should not report on custom element with explicit aria-label attribute", async () => {
		expect.assertions(1);
		const markup = /* HTML */ ` <custom-allowed aria-label="foobar"></custom-allowed> `;
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should report error when aria-label is used on invalid element", async () => {
		expect.assertions(2);
		const markup = /* HTML */ ` <p aria-label="foobar"></p> `;
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeInvalid();
		expect(report).toMatchInlineCodeframe(`
			"error: "aria-label" cannot be used on this element (aria-label-misuse) at inline:1:5:
			> 1 |  <p aria-label="foobar"></p>
			    |     ^^^^^^^^^^
			Selector: p"
		`);
	});

	it("should report error when aria-label is used on input hidden", async () => {
		expect.assertions(2);
		const markup = /* HTML */ ` <input type="hidden" aria-label="foobar" /> `;
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeInvalid();
		expect(report).toMatchInlineCodeframe(`
			"error: "aria-label" cannot be used on this element (aria-label-misuse) at inline:1:23:
			> 1 |  <input type="hidden" aria-label="foobar" />
			    |                       ^^^^^^^^^^
			Selector: input"
		`);
	});

	it("should report error on custom element with without explicit aria-label attribute", async () => {
		expect.assertions(2);
		const markup = /* HTML */ ` <custom-disallowed aria-label="foobar"></custom-disallowed> `;
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeInvalid();
		expect(report).toMatchInlineCodeframe(`
			"error: "aria-label" cannot be used on this element (aria-label-misuse) at inline:1:21:
			> 1 |  <custom-disallowed aria-label="foobar"></custom-disallowed>
			    |                     ^^^^^^^^^^
			Selector: custom-disallowed"
		`);
	});

	it("should handle dynamic attribute", async () => {
		expect.assertions(2);
		const markup = /* HTML */ ` <p dynamic-aria-label="foobar"></p> `;
		const report = await htmlvalidate.validateString(markup, { processAttribute });
		expect(report).toBeInvalid();
		expect(report).toMatchInlineCodeframe(`
			"error: "aria-label" cannot be used on this element (aria-label-misuse) at inline:1:5:
			> 1 |  <p dynamic-aria-label="foobar"></p>
			    |     ^^^^^^^^^^^^^^^^^^
			Selector: p"
		`);
	});

	it("should handle interpolated attribute", async () => {
		expect.assertions(2);
		const markup = /* HTML */ ` <p aria-label="{{ interpolated }}"></p> `;
		const report = await htmlvalidate.validateString(markup, { processAttribute });
		expect(report).toBeInvalid();
		expect(report).toMatchInlineCodeframe(`
			"error: "aria-label" cannot be used on this element (aria-label-misuse) at inline:1:5:
			> 1 |  <p aria-label="{{ interpolated }}"></p>
			    |     ^^^^^^^^^^
			Selector: p"
		`);
	});

	describe("with allowAnyNamable option", () => {
		it("should not report error for elements which allow naming", async () => {
			expect.assertions(1);
			const htmlvalidate = new HtmlValidate({
				elements: ["html5"],
				rules: { "aria-label-misuse": ["error", { allowAnyNamable: true }] },
			});
			const markup = /* HTML */ ` <h1 aria-label="lorem ipsum">spam</h1> `;
			const report = await htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should report error for elements where naming is prohibited", async () => {
			expect.assertions(2);
			const htmlvalidate = new HtmlValidate({
				elements: ["html5"],
				rules: { "aria-label-misuse": ["error", { allowAnyNamable: true }] },
			});
			const markup = /* HTML */ ` <span aria-label="lorem ipsum">spam</span> `;
			const report = await htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toMatchInlineCodeframe(`
				"error: "aria-label" cannot be used on this element (aria-label-misuse) at inline:1:8:
				> 1 |  <span aria-label="lorem ipsum">spam</span>
				    |        ^^^^^^^^^^
				Selector: span"
			`);
		});
	});

	describe("without allowAnyNamable option", () => {
		it("should report error for elements where naming allowed is not recommended", async () => {
			expect.assertions(2);
			const htmlvalidate = new HtmlValidate({
				elements: ["html5"],
				rules: { "aria-label-misuse": ["error", { allowAnyNamable: false }] },
			});
			const markup = /* HTML */ ` <h1 aria-label="lorem ipsum">spam</h1> `;
			const report = await htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toMatchInlineCodeframe(`
				"error: "aria-label" cannot be used on this element (aria-label-misuse) at inline:1:6:
				> 1 |  <h1 aria-label="lorem ipsum">spam</h1>
				    |      ^^^^^^^^^^
				Selector: h1"
			`);
		});
	});

	it("should contain documentation", async () => {
		expect.assertions(1);
		/* eslint-disable-next-line @typescript-eslint/no-deprecated -- technical debt */
		const docs = await htmlvalidate.getRuleDocumentation("aria-label-misuse");
		expect(docs).toMatchSnapshot();
	});
});
