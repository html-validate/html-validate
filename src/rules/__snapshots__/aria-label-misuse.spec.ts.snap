// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule aria-label-misuse should contain documentation 1`] = `
{
  "description": "\`aria-label\` can only be used on:

- Interactive elements
- Labelable elements
- Landmark elements
- Elements with roles inheriting from widget
- \`<area>\`
- \`<form>\` and \`<fieldset>\`
- \`<iframe>\`
- \`<img>\` and \`<figure>\`
- \`<summary>\`
- \`<table>\`, \`<td>\` and \`<th>\`
",
  "url": "https://html-validate.org/rules/aria-label-misuse.html",
}
`;
